### ISSUE TITLE
Create /etc/role if it does not exist

### ISSUE THREAD

**Issue by mikhailnov** [Etersoft/libnss-role#5]
It would be good if `roleadd` created the file `/etc/role` in case it does not exist.

https://github.com/Etersoft/libnss-role/blob/sisyphus/lock_file.c#L37

```strace
# strace roleadd users cdrom
execve("/usr/sbin/roleadd", ["roleadd", "users", "cdrom"], 0x7ffe3a827280 /* 55 vars */) = 0
<...>
open("/etc/role", O_RDONLY)             = -1 ENOENT (Нет такого файла или каталога)
write(2, "Error 153: I/O file error\n", 26Error 153: I/O file error
) = 26
exit_group(153)                         = ?
+++ exited with 153 +++

```

**Comment by mikhailnov** [comment]
Вот этот  код

```cpp
    fd = open(file, O_CREAT | O_EXCL | O_WRONLY, 0600);
    if (-1 == fd) {
        return LIBROLE_OK;
    }
```
(https://github.com/Etersoft/libnss-role/blob/sisyphus/lock_file.c#L45)
вроде бы должен создавать файл, если он отсутствует, но этого не происходит.

**Comment by vpashka** [comment]
Если файл уже существует то возвращается -1 и errno выставляется в EEXIST.
А здесь как я понимаю нет проверки фактической ошибки. Может файл создать не смогли
из-за нехватки прав.

**Comment by mikhailnov** [comment]
Прав достаточно. По выводу strace выше видно, что до приведенного мною в комментарии №2 кода даже не доходит.

Скорее всего, спотыкается в parser.c на
```cpp
f = fopen(s, "r");
    if (!f)
        return LIBROLE_IO_ERROR;
```

**Comment by mikhailnov** [comment]
Хотя в strace open(), а не fopen()

**Comment by vpashka** [comment]
В любом случае, я бы считал этот код не очень корректным (возврат OK), если его цель создать файл, если не существует. Всё-таки Ok нужно возвращать только в случае EEXIST.

**Comment by mikhailnov** [comment]
Да, ошибка происходит здесь: https://github.com/Etersoft/libnss-role/blob/sisyphus/parser.c#L257

Сделал так:
```diff
--- a/parser.c
+++ b/parser.c
@@ -256,6 +256,9 @@ int librole_reading(const char *s, struct librole_graph *G)
 
     f = fopen(s, "r");
     if (!f)
+#ifdef DEBUG
+        printf("Error performing fopen\n");
+#endif
         return LIBROLE_IO_ERROR;
     
     while(1) {
```

`scons -Q DEBUG=yes`

```
$ sudo LD_LIBRARY_PATH=$PWD ./roleadd users cdrom
Error performing fopen
Error 153: I/O file error
```

**Comment by mikhailnov** [comment]
Похоже, что при выполнении `roleadd` сначала запускается парсер, который выдает ошибку, если файла `/etc/role` нет, а нужно бы, если выполняется именно добавление роли, создавать этот файл, при чем создавать файл из `parser.c` не очень красиво.

**Comment by vitlav** [comment]
Вы немного запутались, lock_file занимается ограждением файла /etc/role от параллельного изменения.
```
     if (!f)
+#ifdef DEBUG
+        printf("Error performing fopen\n");
+#endif
         return LIBROLE_IO_ERROR;
```
Тут без фигурных скобок добавление printf приводит к постоянному возврату ошибки.

Какой смысл создавать пустой файл /etc/role? Предполагается, что этот файл приезжает с пакетом.

**Comment by mikhailnov** [comment]
> Тут без фигурных скобок добавление printf приводит к постоянному возврату ошибки.

Точно, спасибо

> Какой смысл создавать пустой файл /etc/role?

Чтобы не было не очень понятных пользователю ошибок, если его нет, но не критично.

**Comment by vitlav** [comment]
Я обнаружил критическую ошибку при отсутствии файла. По возможности никуда не внедряйте эту версию, переписываю код :)

**Comment by ghost** [comment]
I suggest to avoid automatic creation of `/etc/role`. Let's just suppose that missing file is an empty file. User must be the one who decides if the file is needed at all.
